### PR TITLE
Fix support for containment paths in nav prop bindings

### PR DESCRIPTION
**Pull request by mikepizzo**
### Description
Fixed numerous issues with containment nav prop bindings.
    -Fixed issues with locating the proper contained entity set in FindNavigationTarget
    -Fixed nav prop binding issues with nav props on complex types
    -Fixed issues with reading/writing nav prop bindings for contained entity sets
    -Added support for writing context for contained nav props under LibraryCompatibility.Version6

Added support for entity types used exclusively in singletons and single-valued nav props without defining a key.

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### CODE REVIEW

File: test/FunctionalTests/Tests/TestUtils/Common/Microsoft.Test.OData.Utils/ODataLibTest/ModelBuilder.cs [hunk 1]
```diff
@@ -497,7 +497,7 @@ public static void AddDefaultContainerFixup(this EdmModel model, string namespac
                 };
 
             // create NavigationPropertyBinding
-            foreach (EdmNavigationProperty property in model.SchemaElements.OfType<IEdmEntityType>().SelectMany(entityType => entityType.DeclaredNavigationProperties()))
+            foreach (EdmNavigationProperty property in model.SchemaElements.OfType<IEdmEntityType>().SelectMany(entityType => entityType.DeclaredNavigationProperties()).Where(p=>p.ContainsTarget == false))
```
Comment (xuzhg) [inline test/FunctionalTests/Tests/TestUtils/Common/Microsoft.Test.OData.Utils/ODataLibTest/ModelBuilder.cs]: >p=>p.ContainsTarget == false [](start = 175, length = 28)

can use p => !p.ContainsTarget

File: test/FunctionalTests/Tests/DataServices/UnitTests/ServerUnitTests/Stubs/ExpressionTreeTests/ExpressionTreeToXmlSerializer.cs [hunk 2]
```diff
@@ -63,7 +63,7 @@ internal class ExpressionTreeToXmlSerializer : ExpressionVisitor
         public static String SerializeToString(Expression expression)
         {
             outputText = new StringBuilder();
-            writer = XmlTextWriter.Create(outputText);
+            writer = XmlWriter.Create(outputText);
```
Comment (xuzhg) [inline test/FunctionalTests/Tests/DataServices/UnitTests/ServerUnitTests/Stubs/ExpressionTreeTests/ExpressionTreeToXmlSerializer.cs]: >XmlWriter [](start = 21, length = 9)

Is it a bug?

File: test/FunctionalTests/Tests/DataEdmLib/FunctionalTests/VocabularyValidationTests.cs [hunk 3]
```diff
@@ -425,6 +423,151 @@ public void ValidateOutOfLineAnnotationWithValidTargetsInReferencedModel()
             Assert.IsTrue(validationResult, "Expected no validation errors from annotation with targets in an external model.");
         }
 
+        [TestMethod]
+        public void ValidateUnresolvedInlineAnnotationTargets()
```
Comment (xuzhg) [inline test/FunctionalTests/Tests/DataEdmLib/FunctionalTests/VocabularyValidationTests.cs]: >public void ValidateUnresolvedInlineAnnotationTargets() [](start = 8, length = 55)

The test cases seem not related to nav prop bindings?

File: test/FunctionalTests/Tests/DataEdmLib/FunctionalTests/VocabularyValidationTests.cs [hunk 4]
```diff
@@ -425,6 +423,151 @@ public void ValidateOutOfLineAnnotationWithValidTargetsInReferencedModel()
...
+    <Property Name=""ID"" Type=""Edm.String"" Nullable=""false"">
+      <Annotation Term=""NS.CommonValueTerm"" String=""Hello world!"" />
+      <Annotation Term=""AnnotationNS.UnknownValueTerm"" String=""Hello world!"" />
+      <Annotation Term=""RefNS.UnknownValueTerm"" String=""Hello world!""/>
+    </Property>
+    <Annotation Term=""NS.CommonValueTerm"" String=""Hello world!"" />
+    <Annotation Term=""AnnotationNS.UnknownValueTerm"" String=""Hello world!"" />
+    <Annotation Term=""RefNS.UnknownValueTerm"" String=""Hello world!""/>
+  </EntityType>
+  <Action Name=""SomeFunction""><ReturnType Type=""Edm.Int32""/>
```
Comment (xuzhg) [inline test/FunctionalTests/Tests/DataEdmLib/FunctionalTests/VocabularyValidationTests.cs]: ><ReturnType Type=""Edm.Int32""/> [](start = 32, length = 32)

put returntype at new line?

File: test/FunctionalTests/Microsoft.OData.Edm.Tests/Validation/ValidationRuleSetTests.cs [hunk 5]
```diff
@@ -22,7 +22,9 @@ public class ValidationRuleSetTests
         public void NewValidationRulesShouldBeInTheRuleSetExceptBaselinedExceptionRules()
         {
             var validationRules = new Dictionary<object, string>();
-            var items = typeof(ValidationRules).GetFields().Where(f => f.Name != "NavigationPropertyEntityMustNotIndirectlyContainItself")
+            var items = typeof(ValidationRules).GetFields().Where(f =>
+                f.Name != "NavigationPropertyEntityMustNotIndirectlyContainItself" &&
+                f.Name != "EntityTypeKeyMissingOnEntityType")
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Edm.Tests/Validation/ValidationRuleSetTests.cs]: >"EntityTypeKeyMissingOnEntityType" [](start = 26, length = 34)

I'd like to keep it unremoved and add the key in the corresponding test cases?

File: test/FunctionalTests/Microsoft.OData.Edm.Tests/Schema/EdmSingletonTests.cs [hunk 6]
```diff
@@ -9,6 +9,8 @@
 using Microsoft.OData.Edm;
 using Microsoft.OData.Edm.Vocabularies;
 using Xunit;
+using System;
+using System.IO;
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Edm.Tests/Schema/EdmSingletonTests.cs]: move to top

File: test/FunctionalTests/Microsoft.OData.Edm.Tests/ScenarioTests/OasisRelationshipChangesAcceptanceTests.cs [hunk 7]
```diff
@@ -13,6 +13,7 @@
 using Microsoft.OData.Edm.Validation;
 using Xunit;
 using ErrorStrings = Microsoft.OData.Edm.Strings;
+using System.Linq;
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Edm.Tests/ScenarioTests/OasisRelationshipChangesAcceptanceTests.cs]: >System.Linq; [](start = 6, length = 12)

move to top

File: test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlReaderTests.cs [hunk 8]
```diff
@@ -17,6 +17,7 @@
 using Microsoft.OData.Edm.Vocabularies.V1;
 using Xunit;
 using ErrorStrings = Microsoft.OData.Edm.Strings;
+using System.IO;
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Edm.Tests/Csdl/CsdlReaderTests.cs]: >System.IO; [](start = 6, length = 10)

move to top

File: src/Microsoft.OData.Edm/Validation/InterfaceValidator.cs [hunk 9]
```diff
@@ -314,6 +314,22 @@ private IEnumerable<EdmError> ValidateStructure(object item)
...
+                foreach (IEdmVocabularyAnnotation annotation in annotatable.VocabularyAnnotations(this.model))
+                {
+                    // Inline annotations have a null target.
+                    if (annotation.Target == null)
+                    {
+                        // Target must be set for validation, so create a new EdmVocabularyAnnotation with the target set
+                        followupErrors.AddRange(this.ValidateStructure(new EdmVocabularyAnnotation(annotatable, annotation.Term, annotation.Qualifier, annotation.Value)));
+                    }
+                }
+            }
```
Comment (xuzhg) [inline src/Microsoft.OData.Edm/Validation/InterfaceValidator.cs]: That seems not related to the nav prop binding

File: src/Microsoft.OData.Edm/Validation/InterfaceValidator.cs [hunk 10]
```diff
@@ -314,6 +314,22 @@ private IEnumerable<EdmError> ValidateStructure(object item)
...
 
+            // Out-of-line annotations are found through the model visitor, but inline annotations
+            // are not found through normal traversal, so add in-line annotations here.
+            IEdmVocabularyAnnotatable annotatable = item as IEdmVocabularyAnnotatable;
+            if (this.model != null && annotatable != null)
+            {
+                foreach (IEdmVocabularyAnnotation annotation in annotatable.VocabularyAnnotations(this.model))
+                {
+                    // Inline annotations have a null target.
+                    if (annotation.Target == null)
```
Comment (xuzhg) [inline src/Microsoft.OData.Edm/Validation/InterfaceValidator.cs]: >annotation.Target == null [](start = 24, length = 25)

So, do we use "annotation.Target == null" to distinguish inline from outOfLine?

File: src/Microsoft.OData.Edm/Validation/ValidationRuleSet.cs [hunk 11]
```diff
@@ -34,7 +34,7 @@ public sealed class ValidationRuleSet : IEnumerable<ValidationRule>
                 ValidationRules.EntityTypeInvalidKeyNullablePart,
                 ValidationRules.EntityTypeEntityKeyMustBeScalar,
                 ValidationRules.EntityTypeInvalidKeyKeyDefinedInBaseClass,
-                ValidationRules.EntityTypeKeyMissingOnEntityType,
+////                ValidationRules.EntityTypeKeyMissingOnEntityType,
```
Comment (xuzhg) [inline src/Microsoft.OData.Edm/Validation/ValidationRuleSet.cs]: >ValidationRules.EntityTypeKeyMissingOnEntityType, [](start = 20, length = 49)

remove?

File: test/FunctionalTests/Microsoft.OData.Core.Tests/IntegrationTests/Evaluation/AutoComputePayloadMetadataInJsonIntegrationTests.cs [hunk 12]
```diff
@@ -936,6 +936,7 @@ public void WritingInFullMetadataModeWithExpandAndProjectionShouldWriteMissingMe
 
             const string selectClause = "StreamProp1,Namespace.AlwaysBindableAction1,Namespace.AlwaysBindableFunction1,DeferredNavLink";
             const string expandClause = "ExpandedNavLink($select=StreamProp1,Namespace.AlwaysBindableAction1;$expand=ExpandedNavLink($select=StreamProp2,Namespace.AlwaysBindableAction1))";
+
             this.GetWriterOutputForContentTypeAndKnobValue("application/json;odata.metadata=full", true, itemsToWrite, Model, EntitySet, EntityType, selectClause, expandClause)
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/IntegrationTests/Evaluation/AutoComputePayloadMetadataInJsonIntegrationTests.cs]: maybe keep this file unchanged?

File: src/Microsoft.OData.Edm/ExtensionMethods/ExtensionMethods.cs [hunk 13]
```diff
@@ -2735,6 +2735,12 @@ internal static bool TryGetRelativeEntitySetPath(IEdmElement element, Collection
 
                     navigationProperties[navigationProperty] = new EdmPathExpression(paths);
 
+                    // In 7.4.1, FindNavigationTarget expected a binding path that included the path
+                    // to the contained entity set. In 7.4.2 FindNavigationTarget was fixed to work off
+                    // of the path from the contained entity set, but retained the old behavior as well
+                    // for backward compatibility. In the next breaking change we should remove that
+                    // behavior in FindNavigationTarget and remove this special handling of containsTarget
+                    // by always clearing the path.
```
Comment (AlanWong-MS) [inline src/Microsoft.OData.Edm/ExtensionMethods/ExtensionMethods.cs]: Do we have any items to track this task?

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs [hunk 14]
```diff
@@ -786,13 +786,15 @@ protected override void StartNestedResourceInfoWithContent(ODataNestedResourceIn
             {
                 // Write @odata.context annotation for navigation property
                 var containedEntitySet = this.CurrentScope.NavigationSource as IEdmContainedEntitySet;
-                if (containedEntitySet != null)
+                if (containedEntitySet != null && this.jsonLightOutputContext.MessageWriterSettings.LibraryCompatibility == ODataLibraryCompatibility.Version6)
```
Comment (biaol-odata) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs]: >ODataLibraryCompatibility [](start = 124, length = 25)

what should we do for the case of setting.LibraryCompatiblity being Version7 or Latest?

File: src/Microsoft.OData.Core/ODataContextUrlInfo.cs [hunk 15]
```diff
@@ -407,8 +413,13 @@ private static string ProcessSubExpand(string expandNode, string subExpand)
         private static string CombineSelectAndExpandResult(IList<string> selectList, IList<string> expandList)
         {
             string currentExpandClause = string.Empty;
+
             if (selectList.Any())
             {
+                // https://github.com/OData/odata.net/issues/1104
+                // If the user explicitly selects and expands a nav prop, we should include both forms in contextUrl
+                // We can't, though, because SelectExpandBinder.AddExplicitNavPropsWhereNeeded adds all of the expanded items
```
Comment (biaol-odata) [inline src/Microsoft.OData.Core/ODataContextUrlInfo.cs]: >SelectExpandBinder.AddExplicitNavPropsWhereNeeded [](start = 45, length = 49)

This method doesn't exist. Not mentioned in PR #1104 either. Should it be SelectExpandClauseFinisher.AddExplicitNavPropLinksWhereNecessary instead?

File: src/Microsoft.OData.Core/ODataContextUrlInfo.cs [hunk 16]
```diff
@@ -394,10 +399,11 @@ private static string CreateSelectExpandContextUriSegment(SelectExpandClause sel
...
         /// <param name="expandNode">The current expanded node.</param>
         /// <param name="subExpand">Generated sub expand node.</param>
+        /// <param name="version">OData Version of the generated response.</param>
         /// <returns>The generated expand string.</returns>
-        private static string ProcessSubExpand(string expandNode, string subExpand)
+        private static string ProcessSubExpand(string expandNode, string subExpand, ODataVersion version)
         {
-            return string.IsNullOrEmpty(subExpand) ? null : expandNode + ODataConstants.ContextUriProjectionStart + subExpand + ODataConstants.ContextUriProjectionEnd;
+            return string.IsNullOrEmpty(subExpand) && version <= ODataVersion.V4 ? null : expandNode + ODataConstants.ContextUriProjectionStart + subExpand + ODataConstants.ContextUriProjectionEnd;
         }
```
Comment (biaol-odata) [inline src/Microsoft.OData.Core/ODataContextUrlInfo.cs]: nit: please break this long line into two.

File: src/Microsoft.OData.Core/SelectedPropertiesNode.cs [hunk 17]
```diff
@@ -582,19 +583,21 @@ private bool IsOperationSelectedAtThisLevel(IEdmOperation operation, bool mustBe
...
             return node;
         }
 
         /// <summary>Process sub expand node, set name for the node.</summary>
         /// <param name="nodeName">Node name for the subexpandnode.</param>
         /// <param name="subExpandNode">Generated sub expand node.</param>
+        /// <param name="version">OData version.</param>
         /// <returns>The sub expanded node passed in.</returns>
-        private static SelectedPropertiesNode ProcessSubExpand(string nodeName, SelectedPropertiesNode subExpandNode)
+        private static SelectedPropertiesNode ProcessSubExpand(string nodeName, SelectedPropertiesNode subExpandNode, ODataVersion version)
```
Comment (biaol-odata) [inline src/Microsoft.OData.Core/SelectedPropertiesNode.cs]: >, ODataVersion version [](start = 116, length = 22)

This newly added "version" argument is not referenced/used in this function?

File: src/Microsoft.OData.Edm/EdmUtil.cs [hunk 18]
```diff
@@ -231,7 +231,11 @@ internal static string ParameterizedName(IEdmOperation operation)
             foreach (IEdmOperationParameter parameter in operation.Parameters)
             {
                 string typeName = "";
-                if (parameter.Type.IsCollection())
+                if (parameter.Type == null)
```
Comment (biaol-odata) [inline src/Microsoft.OData.Edm/EdmUtil.cs]: >parameter.Type == null [](start = 20, length = 22)

I cannot see how this typeName resolution for Untyped related to nav prop binding. Is this for a separate issue?

File: src/Microsoft.OData.Edm/Schema/EdmContainedEntitySet.cs [hunk 19]
```diff
@@ -7,37 +7,52 @@
...
 {
     using System.Collections.Generic;
+    using System.Linq;
 
     /// <summary>
     /// Represents an EDM contained entity set.
     /// </summary>
     internal class EdmContainedEntitySet : EdmEntitySetBase, IEdmContainedEntitySet
     {
+        internal readonly IEdmPathExpression NavigationPath;
```
Comment (biaol-odata) [inline src/Microsoft.OData.Edm/Schema/EdmContainedEntitySet.cs]: >NavigationPath [](start = 45, length = 14)

nit: should we use a private member named "navigationPath", and expose it thru internal property named "NavigationPath"?

File: test/FunctionalTests/Framework/Workspaces/NonClrWorkspaces/NonClrService/NonClrExpressionTreeToXmlSerializer.cs [hunk 20]
```diff
@@ -69,7 +69,7 @@ internal class ExpressionTreeToXmlSerializer : System.Data.Test.Astoria.LateBoun
         public static String Serialize(Expression exp)
         {
             outputText = new StringBuilder();
-            writer = XmlTextWriter.Create(outputText);
+            writer = XmlWriter.Create(outputText);
```
Comment (biaol-odata) [inline test/FunctionalTests/Framework/Workspaces/NonClrWorkspaces/NonClrService/NonClrExpressionTreeToXmlSerializer.cs]: >XmlWriter.Create(outputText); [](start = 21, length = 29)

doesn't seem like this change is related to nav prop bindings. Code expression simplification instead?

File: test/FunctionalTests/Tests/DataEdmLib/FunctionalTests/VocabularyValidationTests.cs (hunk 3 above)
Comment (mikepizzo) [reply test/FunctionalTests/Tests/DataEdmLib/FunctionalTests/VocabularyValidationTests.cs]: Yes; I ran across several other issues in trying to fix navpropbindings, and in order to get a build out that included all of the fixes I put them in the same PR.

---
In reply to: [174972862](https://github.com/OData/odata.net/pull/1109#discussion_r174972862) [](ancestors = 174972862)

File: src/Microsoft.OData.Edm/Validation/InterfaceValidator.cs (hunk 10 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Edm/Validation/InterfaceValidator.cs]: Yes.

---
In reply to: [174975029](https://github.com/OData/odata.net/pull/1109#discussion_r174975029) [](ancestors = 174975029)

File: src/Microsoft.OData.Edm/Validation/InterfaceValidator.cs (hunk 9 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Edm/Validation/InterfaceValidator.cs]: Right; I ran into this while working through some related issues and just fixed it here rather than opening a separate PR.

---
In reply to: [174974871](https://github.com/OData/odata.net/pull/1109#discussion_r174974871) [](ancestors = 174974871)

File: src/Microsoft.OData.Edm/Validation/ValidationRuleSet.cs (hunk 11 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Edm/Validation/ValidationRuleSet.cs]: I wanted to leave it here, explicitly commented out, so we knew we intended it to be disabled.

---
In reply to: [174975195](https://github.com/OData/odata.net/pull/1109#discussion_r174975195) [](ancestors = 174975195)

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs (hunk 14 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs]: We shouldn't write context Url for nav props unless the contextUrl doesn't follow convention; and we currently have no way for the app to tell us that it doesn't follow convention, so we should just not write it.

---
In reply to: [175302067](https://github.com/OData/odata.net/pull/1109#discussion_r175302067) [](ancestors = 175302067)

File: src/Microsoft.OData.Edm/EdmUtil.cs (hunk 18 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Edm/EdmUtil.cs]: Yes. I combined into same PR so we could verify in the same nightly builds.

---
In reply to: [175302690](https://github.com/OData/odata.net/pull/1109#discussion_r175302690) [](ancestors = 175302690)

File: test/FunctionalTests/Framework/Workspaces/NonClrWorkspaces/NonClrService/NonClrExpressionTreeToXmlSerializer.cs (hunk 20 above)
Comment (mikepizzo) [reply test/FunctionalTests/Framework/Workspaces/NonClrWorkspaces/NonClrService/NonClrExpressionTreeToXmlSerializer.cs]: This was a bug that prevent compilation on certain platforms.

---
In reply to: [175303989](https://github.com/OData/odata.net/pull/1109#discussion_r175303989) [](ancestors = 175303989)

File: src/Microsoft.OData.Core/ODataContextUrlInfo.cs (hunk 15 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Core/ODataContextUrlInfo.cs]: Yes; thanks.

---
In reply to: [175302307](https://github.com/OData/odata.net/pull/1109#discussion_r175302307) [](ancestors = 175302307)

File: src/Microsoft.OData.Core/SelectedPropertiesNode.cs (hunk 17 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Core/SelectedPropertiesNode.cs]: No, but we want to pass this function as the first parameter to SelectExpandClause.Traverse, which takes a Func<string, SelectedPropertiesNode, ODataVersion); it's not used in this case, but it is used in other cases so it needs to be part of the signature.

---
In reply to: [175302563](https://github.com/OData/odata.net/pull/1109#discussion_r175302563) [](ancestors = 175302563)

File: test/FunctionalTests/Microsoft.OData.Edm.Tests/Validation/ValidationRuleSetTests.cs (hunk 5 above)
Comment (mikepizzo) [reply test/FunctionalTests/Microsoft.OData.Edm.Tests/Validation/ValidationRuleSetTests.cs]: We can't remove it, because it is public, but we don't want it to be part of the ValidationRules.  This test verifies that all of the defined tests are part of the validation rules, so we need to omit the two that are defined but should not be part of the validation rules.

---
In reply to: [174973472](https://github.com/OData/odata.net/pull/1109#discussion_r174973472) [](ancestors = 174973472)

File: src/Microsoft.OData.Edm/ExtensionMethods/ExtensionMethods.cs (hunk 13 above)
Comment (mikepizzo) [reply src/Microsoft.OData.Edm/ExtensionMethods/ExtensionMethods.cs]: Opened [https://github.com/OData/odata.net/issues/1114](https://github.com/OData/odata.net/issues/1114).

---
In reply to: [175176585](https://github.com/OData/odata.net/pull/1109#discussion_r175176585) [](ancestors = 175176585)

File: test/FunctionalTests/Tests/DataServices/UnitTests/ServerUnitTests/Stubs/ExpressionTreeTests/ExpressionTreeToXmlSerializer.cs (hunk 2 above)
Comment (mikepizzo) [reply test/FunctionalTests/Tests/DataServices/UnitTests/ServerUnitTests/Stubs/ExpressionTreeTests/ExpressionTreeToXmlSerializer.cs]: XmlTextWriter.Create is not supported in some environments.

---
In reply to: [174972609](https://github.com/OData/odata.net/pull/1109#discussion_r174972609) [](ancestors = 174972609)
